### PR TITLE
Remove owner references from NetworkAddonsConfig on create/update.

### PR DESCRIPTION
**Pull request by ormergi**
NetworkAddonsConfigs CR is cluster scoped, but has owner reference to HyperConverge CR which is namespace scoped.
Thus NetworkAddonsConfigs will be removed eventualy by the kubernetes garbage-collector.

On NetworkAddonsConfigs creation do not set any owner refrence.
On update, checks if there are owner refrences and remove them.

### CODE REVIEW

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

**Comment by dankenigsberg** [comment]
@ormergi please add a word on what happens when the kind-HCO is being deleted?

**Comment by djzager** [comment]
> @ormergi please add a word on what happens when the kind-HCO is being deleted?

nothing will happen to the NetworkAddonsConfig resource. it will remain. That is what I believe to be problematic of this and #393 being done before #380 is merged.

File: pkg/controller/hyperconverged/hyperconverged_controller.go [hunk 1]
```diff
@@ -616,6 +612,14 @@ func (r *ReconcileHyperConverged) ensureNetworkAddons(instance *hcov1alpha1.Hype
 
 	logger.Info("NetworkAddonsConfig already exists", "NetworkAddonsConfig.Namespace", found.Namespace, "NetworkAddonsConfig.Name", found.Name)
 
+	existingOwners := found.GetOwnerReferences()
+	if len(existingOwners) > 0 {
+		logger.Info("Removing all NetworkAddonsConfig owner references","NetworkAddonsConfig.OwnerReferences", found.GetOwnerReferences())
+		emptyOwnerReferencesArray := []metav1.OwnerReference{}
+		found.SetOwnerReferences(emptyOwnerReferencesArray)
```
Comment (djzager) [inline pkg/controller/hyperconverged/hyperconverged_controller.go]:
```suggestion
		found.SetOwnerReferences([]metav1.OwnerReference{})
```

File: pkg/controller/hyperconverged/hyperconverged_controller.go [hunk 2]
```diff
@@ -616,6 +612,14 @@ func (r *ReconcileHyperConverged) ensureNetworkAddons(instance *hcov1alpha1.Hype
 
 	logger.Info("NetworkAddonsConfig already exists", "NetworkAddonsConfig.Namespace", found.Namespace, "NetworkAddonsConfig.Name", found.Name)
 
+	existingOwners := found.GetOwnerReferences()
+	if len(existingOwners) > 0 {
+		logger.Info("Removing all NetworkAddonsConfig owner references","NetworkAddonsConfig.OwnerReferences", found.GetOwnerReferences())
+		emptyOwnerReferencesArray := []metav1.OwnerReference{}
```
Comment (djzager) [inline pkg/controller/hyperconverged/hyperconverged_controller.go]:
```suggestion
```

File: pkg/controller/hyperconverged/hyperconverged_controller.go (hunk 1 above)
Comment (ormergi) [reply pkg/controller/hyperconverged/hyperconverged_controller.go]: done

File: pkg/controller/hyperconverged/hyperconverged_controller.go (hunk 2 above)
Comment (ormergi) [reply pkg/controller/hyperconverged/hyperconverged_controller.go]: done

**Comment by ormergi** [comment]
@dankenigsberg 
Once the HCO is deleted  NetworkAddonsConfigs will remain, another effect for removing HCO ownerRefrence is that HCO is not a subscriber of NetworkAddonsConfigs events anymore and will not handle those.

I am working on adding a finalizer so the CR will be deleted, and understanding which events the hco need to handle then adding a watch to NetworkAddonsConfigs events in a different way from what it is now (by ownerReference)

@djzager 
can you please explain why merging 380 first solves NetworkAddonsConfigs CR deletion?

**Comment by maya-r** [comment]
> @djzager
> can you please explain why merging 380 first solves NetworkAddonsConfigs CR deletion?

As far as I understand...

in https://github.com/kubevirt/hyperconverged-cluster-operator/blob/a3d4cd9ca484d033ad75b5b6d64a7a4d445331d1/pkg/controller/hyperconverged/hyperconverged_controller.go#L87-L99 we make whatever is the owner handler events from this child.  
since HCO is no longer the owner, we no longer run HCO's reconcile whenever an event happens in the child (CNA).

#380 switches to EnqueueRequestsFromMapFunc, it handles events in reconcile based on a map function rather than ownership relationship.  
The map should always say "wake up HCO's reconcile function".

**Comment by dankenigsberg** [comment]
> > @djzager
> > can you please explain why merging 380 first solves NetworkAddonsConfigs CR deletion?
> 
> As far as I understand...
> 
> in
> 
> https://github.com/kubevirt/hyperconverged-cluster-operator/blob/a3d4cd9ca484d033ad75b5b6d64a7a4d445331d1/pkg/controller/hyperconverged/hyperconverged_controller.go#L87-L99
> we make whatever is the owner handler events from this child.
> since HCO is no longer the owner, we no longer run HCO's reconcile whenever an event happens in the child (CNA).
> 
> #380 switches to EnqueueRequestsFromMapFunc, it handles events in reconcile based on a map function rather than ownership relationship.
> The map should always say "wake up HCO's reconcile function".

Yes, I understand why we would like to incorporate something along these lines in order to keep receiving events from the network CR. I do not understand why we'd like to rush to take #380 in its entirety.
Thanks, I got this bit. Now I do not understand why we'd want to drag the rest of #380 with it.

**Comment by ormergi** [comment]
Working on adding a watch to the NetworkAddonsConfigs CR (which has no owner reference)

**Comment by djzager** [comment]
Recommend this be closed in favor of #393

**Comment by djzager** [comment]
Recommend this be closed in favor of #393

**Comment by AlonaKaplan** [comment]
@djzaher why are you recommending to close this pr? #393 is not taking care for removing the owner reference from the networkAddons cr and not taking care for deleting the networkAddons cr in case hco is marked for deletion.  This pr can depend on #393. Or the non-cdi specific code in #393 can be moved to a separate PR and both #395 and #393 will depend on it.

**Comment by djzager** [comment]
> @djzaher why are you recommending to close this pr? #393 is not taking care for removing the owner reference from the networkAddons cr and not taking care for deleting the networkAddons cr in case hco is marked for deletion. This pr can depend on #393. Or the non-cdi specific code in #393 can be moved to a separate PR and both #395 and #393 will depend on it.

- [This](https://github.com/kubevirt/hyperconverged-cluster-operator/pull/393/files#diff-64e1e2b9251fa5137c91d5a7c8427b88R213) is where networkAddons deletion happens as part of handling the finalizer
- [This](https://github.com/kubevirt/hyperconverged-cluster-operator/pull/393/files#diff-64e1e2b9251fa5137c91d5a7c8427b88R666) is where owner references on networkAddons is happening

Am I missing something?

**Comment by djzager** [comment]
Even more than that, @AlonaKaplan, the title of #393 was changes specifically to mention NetworkAddonsConfig. Let me know where I have erred please.

**Comment by AlonaKaplan** [comment]
Thanks for the detailed response. Since #393 was updated to contain network changes, this pr can be closed.

**Comment by djzager** [comment]
> Thanks for the detailed response. Since #393 was updated to contain network changes, this pr can be closed.

I would like to believe that I made the comment after having seen #393 including the changes for network addons but I can't guarantee it. Verifying that this should be closed is absolutely the right thing to do.

Closing this in favor of #393

**Pull request closed without merging**